### DISCUSSION TITLE
Unidirectional CDC option

### DISCUSSION THREAD

**Discussion by bremenpl** [hathach/tinyusb#1923] [Q&A]
Hello,
I am using the CDC device as a logger in my application. I am low on RAM resource and I would like to completely disable CDC RX functionality, as it is not needed in the application. I could not find any dedicated way for doing that. I have noticed that I can reduce the `CFG_TUD_CDC_RX_BUFSIZE`  to a low value (even down to 0), but then the CDC TX parts starts to act in an unexpected way- the sent out data is being stalled until I reset the port (reopen) from the host side. I don't want to go into unnecessary details here, as maybe simply I am not using the CDC code part correctly.

In other words: What should I do/ call, in an application in which the CDC RX part is not needed and I don't want to read any data?
I would appreciate all help!

**Comment by bremenpl** [comment]
As the matter of the fact, weird things start to happen when I set the `CFG_TUD_CDC_RX_BUFSIZE` to less than 128. Once I do that, I notice that the `tud_cdc_rx_cb` is not being called anymore by the tusb stack. Going deeper with the debugger, I have placed a breakpoint in here:

https://github.com/hathach/tinyusb/blob/master/src/class/cdc/cdc_device.c#L435

If the `CFG_TUD_CDC_RX_BUFSIZE >= 128`, it is hit normally when I send a character over CDC. If its less, then this breakpoint is never hit. 

So what I found coincidental, and that it seems to be true, is that the `CFG_TUD_CDC_RX_BUFSIZE` cannot be lower than `CFG_TUD_CDC_EP_BUFSIZE`... When I set both `CFG_TUD_CDC_EP_BUFSIZE` and `CFG_TUD_CDC_RX_BUFSIZE` to 127, then reception worked. Now leaving `CFG_TUD_CDC_EP_BUFSIZE` at 127 and reducing `CFG_TUD_CDC_RX_BUFSIZE` to 126 makes it stop working again. There is no correlation with `CFG_TUD_CDC_TX_BUFSIZE` however. So for example this is OK:

```
// CDC FIFO size of TX and RX
#define CFG_TUD_CDC_RX_BUFSIZE   	(64)
#define CFG_TUD_CDC_TX_BUFSIZE   	(128)

// CDC Endpoint transfer buffer size, more is faster
#define CFG_TUD_CDC_EP_BUFSIZE   	(64)
```

But not this:

```
// CDC FIFO size of TX and RX
#define CFG_TUD_CDC_RX_BUFSIZE   	(63) // <----------
#define CFG_TUD_CDC_TX_BUFSIZE   	(128)

// CDC Endpoint transfer buffer size, more is faster
#define CFG_TUD_CDC_EP_BUFSIZE   	(64) 
```

Going further, I started to wonder whether this line is by design: https://github.com/hathach/tinyusb/blob/master/src/class/cdc/cdc_device.c#L69

Because if I change 'CFG_TUSB_MEM_ALIGN uint8_t epout_buf[CFG_TUD_CDC_EP_BUFSIZE];` to `CFG_TUSB_MEM_ALIGN uint8_t epout_buf[CFG_TUD_CDC_RX_BUFSIZE];` (so the `epout_buf` size is the same as `rx_ff_buf`), I get no issues. It doesn't seem correct that `rx_ff_buf` and `tx_ff_buf` sizes can be different, but not 'epout_buf' and 'epin_buf'.

I would appreciate all help in understanding this...
